mongo/options/clientoptions.go

```diff
@@ -258,328 +256,381 @@ type ClientOptions struct {
 	// may be used in its place to control the amount of time that a single operation can run before returning
 	// an error. Setting SocketTimeout and Timeout on a single client will result in undefined behavior.
 	SocketTimeout *time.Duration
+
+	connString *connstring.ConnString
+}
+
+// ClientOptions contains options to configure a Client instance. Each option
+// can be set through setter functions. See documentation for each setter
+// function for an explanation of the option.
+type ClientOptions struct {
+	Opts []func(*ClientArgs) error
 }
 
 // Client creates a new ClientOptions instance.
 func Client() *ClientOptions {
-	return &ClientOptions{
-		HTTPClient: httputil.DefaultHTTPClient,
-	}
+	opts := &ClientOptions{}
+	opts = opts.SetHTTPClient(httputil.DefaultHTTPClient)
+
+	return opts
 }
 
-// Validate validates the client options. This method will return the first error found.
-func (c *ClientOptions) Validate() error {
-	if c.err != nil {
-		return c.err
-	}
-	c.err = c.validate()
-	return c.err
+// ArgsSetters returns a list of ClientArgs setter functions.
+func (c *ClientOptions) ArgsSetters() []func(*ClientArgs) error {
+	return c.Opts
 }
 
-func (c *ClientOptions) validate() error {
-	// Direct connections cannot be made if multiple hosts are specified or an SRV URI is used.
-	if c.Direct != nil && *c.Direct {
-		if len(c.Hosts) > 1 {
+// ValidateClientArgs checks if the client arguments will create a valid
+// connection.
+func ValidateClientArgs(args *ClientArgs) error {
```

Comment:
It seems like this is separate from `ClientOptions.Validate` so it can also be called in the `topology` package for a `ClientArgs`. Can we move the `Validate` method to a `ClientArgs` instead so we don't have to add this exported symbol?

Alternatively, we should move this validation to an internal package and remove `ClientOptions.Validate`.